### ISSUE TITLE
live demos not working on Safari or Chrome IOS

### ISSUE THREAD

**Issue by gavan1** [alemart/martins-js#4]
Video stream freezes and motion from device motion not being pickup up. 

Feature tracking is working and 3d objects are appearing.#3

**Comment by skeegan** [comment]
I'm seeing the same result. This is the error-

martins.min.js:20 [martins-js] Tracking error: o: WebGL error. 

---------- ERROR ----------

0:73: S0032: no default precision defined for variable 'vec4[9]'
Program has not been successfully linked.


---------- SOURCE CODE ----------

  1. #version 300 es
  2. 
  3. #if 0 == 0
  4. precision mediump float; 
  5. precision mediump sampler2D;
  6. precision highp int; 
  7. #endif
  8. 
  9. #if 0 == 0
 10. #define OUT_TYPE mediump vec4
 11. #elif 0 == 1
 12. #define OUT_TYPE mediump ivec4
 13. #elif 0 == 2
 14. #define OUT_TYPE mediump uvec4
 15. #else
 16. #error Unknown FS_OUTPUT_TYPE
 17. #endif
 18. 
 19. out OUT_TYPE color;
 20. in highp vec2 texCoord;
 21. uniform highp vec2 texSize;
 22. #ifndef _GLOBAL_GLSL
 23. #define _GLOBAL_GLSL
 24. #define threadLocation() ivec2(texCoord * texSize)
 25. #define outputSize() ivec2(texSize)
 26. #define threadPixel(img) textureLod((img), texCoord, 0.0f)
 27. #define pixelAt(img, pos) texelFetch((img), (pos), 0)
 28. #define pixelAtShortOffset(img, offset) textureLodOffset((img), texCoord, 0.0f, (offset))
 29. #define pixelAtLongOffset(img, offset) textureLod((img), texCoord + vec2(offset) / texSize, 0.0f)
 30. #endif
 31. #ifndef _PYRAMIDS_GLSL
 32. #define _PYRAMIDS_GLSL
 33. #define pyrPixel(pyr, lod) textureLod((pyr), texCoord, (lod))
 34. #define pyrPixelAtOffset(pyr, lod, pot, offset) textureLod((pyr), texCoord + ((pot) * vec2(offset)) / texSize, (lod))
 35. #define pyrPixelAt(pyr, pos, lod) textureLod((pyr), (vec2(pos) + vec2(0.5f)) / texSize, (lod))
 36. #define pyrPixelAtEx(pyr, pos, lod, pyrBaseSize) textureLod((pyr), (vec2(pos) + vec2(0.5f)) / vec2(pyrBaseSize), (lod))
 37. #define pyrSubpixelAtEx(pyr, pos, lod, pyrBaseSize) textureLod((pyr), ((pos) + vec2(0.5f)) / vec2(pyrBaseSize), (lod))
 38. #define pyrSubpixelAtExOffset(pyr, pos, lod, pot, offset, pyrBaseSize) textureLod((pyr), (((pos) + vec2(0.5f)) + ((pot) * vec2(offset))) / vec2(pyrBaseSize), (lod))
 39. const int PYRAMID_MAX_LEVELS = int(8);
 40. const float F_PYRAMID_MAX_LEVELS = float(8);
 41. const float LOG2_PYRAMID_MAX_SCALE = float(0);
 42. #define encodeLod(lod) ((LOG2_PYRAMID_MAX_SCALE + (lod)) / (LOG2_PYRAMID_MAX_SCALE + F_PYRAMID_MAX_LEVELS))
 43. float decodeLod(float encodedLod)
 44. {
 45. float lod = encodedLod * (LOG2_PYRAMID_MAX_SCALE + F_PYRAMID_MAX_LEVELS) - LOG2_PYRAMID_MAX_SCALE;
 46. return lod - lod * step(1.0f, encodedLod);
 47. }
 48. #define LOD_EPS 0.0625f
 49. const float ENCODED_LOD_EPS = (LOD_EPS / (LOG2_PYRAMID_MAX_SCALE + F_PYRAMID_MAX_LEVELS));
 50. #define isSameLod(lod1, lod2) (abs((lod1) - (lod2)) < LOD_EPS)
 51. #define isSameEncodedLod(alpha1, alpha2) (abs((alpha1) - (alpha2)) < ENCODED_LOD_EPS)
 52. #endif
 53. #ifndef _FLOAT16_GLSL
 54. #define _FLOAT16_GLSL
 55. #define encodeFloat16(f) (vec2(packf16(f)) / 255.0f)
 56. #define decodeFloat16(v) unpackf16(uvec2((v) * 255.0f))
 57. #define encodePairOfFloat16(f) vec4(encodeFloat16((f).x), encodeFloat16((f).y))
 58. #define decodePairOfFloat16(v) vec2(decodeFloat16((v).rg), decodeFloat16((v).ba))
 59. #define encodeNullPairOfFloat16() vec4(1.0f)
 60. #define isNullPairOfFloat16(v) all(equal((v), encodeNullPairOfFloat16()))
 61. #define encodeDiscardedPairOfFloat16() vec4(0.0f, 1.0f, 0.0f, 1.0f)
 62. #define isDiscardedPairOfFloat16(v) all(equal((v), encodeDiscardedPairOfFloat16()))
 63. #define encodeFloat16NaN() vec2(0.5f, 1.0f)
 64. #define isEncodedFloat16NaN(v) all(equal((v), encodeFloat16NaN()))
 65. uvec2 packf16( float f)
 66. {
 67. uint y = packHalf2x16(vec2(f, 0.0f));
 68. return uvec2(y, y >> 8u) & 0xFFu;
 69. }
 70. float unpackf16(uvec2 v)
 71. {
 72. v &= 0xFFu;
 73. return unpackHalf2x16(v.x | (v.y << 8u)).x;
 74. }
 75. bool isEncodedFloat16Zero(vec2 v)
 76. {
 77. uvec2 w = uvec2(v * 255.0f);
 78. return 0u == w.x + w.y * (0x80u - w.y);
 79. }
 80. #endif
 81. uniform sampler2D corners;
 82. void main()
 83. {
 84. ivec2 thread = threadLocation();
 85. vec4 pixel = threadPixel(corners);
 86. float encodedLod = pixel.a;
 87. float score = decodeFloat16(pixel.rb);
 88. float lod = decodeLod(encodedLod);
 89. float pot = exp2(lod);
 90. color = pixel;
 91. if(score == 0.0f)
 92. return;
 93. #if 1
 94. vec2 gridSize = vec2(pot);
 95. vec2 gridLocation = floor(mod(texCoord * texSize, gridSize));
 96. vec2 gridDelta = gridLocation / gridSize - vec2(0.5f);
 97. float gridStep = 1.0f / pot;
 98. const float adjustment = 1.25f;
 99. color.rb = encodeFloat16(0.0f);
100. if(max(abs(gridDelta.x), abs(gridDelta.y)) > adjustment * gridStep)
101. return;
102. #endif
103. #define P(x,y) textureLod(corners, texCoord + pot * vec2((x), (y)) / texSize, 0.0f)
104. vec4 pix[9] = vec4[9](
105. P(-1,-1), P(0,-1), P(1,-1),
106. P(-1,0), pixel, P(1,0),
107. P(-1,1), P(0,1), P(1,1)
108. );
109. #define S(j) decodeFloat16(pix[j].rb)
110. mat3 scores = mat3(
111. S(0), S(1), S(2),
112. S(3), S(4), S(5),
113. S(6), S(7), S(8)
114. );
115. #define B(j) float(isSameLod(decodeLod(pix[j].a), lod))
116. mat3 sameLod = mat3(
117. B(0), B(1), B(2),
118. B(3), B(4), B(5),
119. B(6), B(7), B(8)
120. );
121. mat3 sameLodScores = matrixCompMult(scores, sameLod);
122. vec3 maxScore3 = max(sameLodScores[0], max(sameLodScores[1], sameLodScores[2]));
123. float maxScore = max(maxScore3.x, max(maxScore3.y, maxScore3.z));
124. color.rb = encodeFloat16(score * step(maxScore, score));
125. }

-> [speedy-vision.js]
warning @ martins.min.js:20

**Comment by alemart** [comment]
@skeegan Thanks for the feedback. I intend to get it working on Safari and iOS.

The error you are seeing has to do with a shader not being compiled due to a driver bug. I fix these kinds of errors in my library [Speedy Vision](https://github.com/alemart/speedy-vision). Once they are fixed, the AR demos here will work.

May I ask you to test the following Speedy Vision demos?

1. https://alemart.github.io/speedy-vision/demos/best-features.html
2. https://alemart.github.io/speedy-vision/demos/feature-matching.html
3. https://alemart.github.io/speedy-vision/demos/feature-matching-lsh.html

You are expected to see points and lines. If you see errors in the console, please let me know.

**Comment by skeegan** [comment]
Hi @alemart , Thanks for your response. Each of those three works for me in Chrome on a Google Pixel 7A (no errors in console).

Any idea on when the other AR demo will be working?

**Comment by alemart** [comment]
@skeegan Thanks for testing. I have updated the Speedy Vision version used in the demos.

May you test the AR demos again? Try the [basic webcam](https://alemart.github.io/martins-js/demo/) and the [WebGL](https://webar-martins-js-hello-webgl.glitch.me/) demos. You are expected to see an animation in augmented reality. If you see any errors in the console, let me know.

You should see MARTINS.js version 0.1.2-wip and Speedy Vision version 0.9.1-wip in the console. Please clear the browser cache if you don't.

**Comment by skeegan** [comment]
Hi @alemart,

I'm seeing the correct versions for MARTINS.js and for Speedy Vision. However, I'm now getting a different error:

martins.min.js:20 [martins-js] Tracking error: r: WebGL error. 

_---------- ERROR ----------

0:207: S0032: no default precision defined for variable 'mat3[2]'
Program has not been successfully linked._

[
[alemart.github.io-1704882393420.log](https://github.com/alemart/martins-js/files/13886362/alemart.github.io-1704882393420.log)
](url)
(full log attached)

**Comment by alemart** [comment]
@skeegan This is a similar driver bug that has to do with an initializer. Would you try again? If you see the same error, please clear the browser cache.

**Comment by alemart** [comment]
> it works now- thanks.

Awesome!

> Quick question: You say that Safari support is coming soon. Any idea when that might be ready?

Yes: when a volunteer helps me test it on Safari. Support for Safari is probably just a tiny change ahead.

Since I started this project back in 2022, the Epiphany browser (WebKit-based) received WebGL2 support. I was able to run the demos on a version of Epiphany.

I don't own a Mac nor an iOS device at the present time, so I'm using Epiphany as a suitable alternative to Safari. I would like to have this project tested on actual Apple hardware as well.

**Comment by skeegan** [comment]
I don't have an iOS device either but I'll check it out when I can. Could
your code be adapted to enable floor detection (for tap-to-place)?

**Comment by alemart** [comment]
> Could your code be adapted to enable floor detection (for tap-to-place)?

That would require implementing a new tracker. I don't have this feature in mind at this moment. I thank you for your interest in my work.

**Comment by alemart** [comment]
To whom it might interest: Safari 15 is the minimum required version.

If you're a Safari user, your feedback is appreciated.

**Comment by alemart** [comment]
I have good news: now we have Safari support.

I managed to test the WebAR demos on several iOS devices. After a small adjustment, I could get them to work on versions 15.2 or newer of the Safari browser (though not on Safari 15.0-15.1).

I have tested the demos on the following devices: iPhone 11 Pro (iOS 15.5), iPhone 12 (iOS 15.6.1), iPhone 12 Pro (iOS 15.0.2), iPhone 13 (iOS 15.0.2), iPhone 13 Pro (iOS 15.2), iPhone 13 Pro Max (iOS 15.1), iPhone 14 (iOS 16.1 and iOS 16.6.1), iPhone XR (iOS 16.4), iPad mini (iPadOS 15.1) and iPad Air (iPadOS 15.4.1). The demos work on every device with Safari 15.2 or newer. Tested on the AWS Device Farm.

If you're a iOS or a macOS user, your feedback is much appreciated. I was not able to test the camera feed on the cloud, and yet the engine seemed to work as expected.

Here is a video of a demo working on iOS (the FPS/GPU rates indicate good performance):

https://github.com/alemart/martins-js/assets/221477/f227bb42-cfdd-4ce4-b013-b6f3cba6ff8e
